### PR TITLE
Handle OAuth2 errors properly

### LINKED ISSUE #2926: Nicer error messages for invalid oauth usage
Currently if anything goes wrong during an oauth flow (no "code" in url, no oauth method set, etc), an exception is thrown.

This is *okay* if the user who stumbles upon it is a site admin that can view the full page, but looks bad to normal users. 

If we used the error flash system to display these errors on the login/register page instead, the experience would be much smoother. 
Not a breaking change since templates should already support errors on login/register page

### PR DESCRIPTION
**Pull request by supercrafter100**
Closes #2926

### CODE REVIEW

**Comment by Aberdeener** [comment]
Could this also handle the “invalid provider” and “no oauth method set” exceptions?

File: modules/Core/pages/oauth.php [hunk 1]
```diff
@@ -2,12 +2,16 @@
...
-        throw new RuntimeException("Invalid provider {$_GET['provider']}");
+        ErrorHandler::logWarning("Invalid provider {$_GET['provider']}");
+        Session::flash('home_error', $language->get('general', 'oauth_failed'));
+        Redirect::to(URL::build('/'));
     }
 }
 
 if (!Session::exists('oauth_method')) {
-    throw new RuntimeException('No OAuth method set');
+    ErrorHandler::logWarning("No OAuth Zmethod set");
```
Comment (partydragen) [inline modules/Core/pages/oauth.php:12]: Why z

Comment (supercrafter100) [reply modules/Core/pages/oauth.php]: shh

File: custom/languages/en_UK.json [hunk 2]
```diff
@@ -780,6 +780,8 @@
     "general/none": "None",
     "general/not_registered_yet": "Not registered yet?",
     "general/oauth_message_continue": "Continue registering with {{provider}}",
+    "general/oauth_failed_setup": "Something went wrong. The OAuth2 setup is invalid. Contact an administrator",
+    "general/oauth_failed": "Something went wrong while trying to login/register/link with your OAuth account. Please try again",
```
Comment (Aberdeener) [inline custom/languages/en_UK.json:784]: "Something went wrong while trying to use OAuth. Please try again."

File: modules/Core/pages/oauth.php [hunk 3]
```diff
@@ -2,19 +2,41 @@
...
-$token = $provider->getAccessToken('authorization_code', [
-    'code' => $_GET['code']
-]);
+try {
+    $token = $provider->getAccessToken('authorization_code', [
+        'code' => $_GET['code']
+    ]);
+} catch (\League\OAuth2\Client\Provider\Exception\IdentityProviderException $e) {
+    Session::flash('oauth_error', $language->get('general', 'oauth_failed_setup'));
+    ErrorHandler::logWarning('An error occurred while handling an oauth ' . Session::get('oauth_method') . ' request: ' . $e->getMessage());
```
Comment (Aberdeener) [inline modules/Core/pages/oauth.php:25]: "OAuth" please 😄

**Review by Aberdeener — CHANGES_REQUESTED** [review]

**Review by Aberdeener — APPROVED** [review]

**Review by samerton — APPROVED** [review]

**Review by partydragen — APPROVED** [review]

**Pull request merged**